### DISCUSSION TITLE
Rails and Restarts

### DISCUSSION THREAD

**Discussion by honzasterba** [puma/puma#3247] [Q&A]
After migrating to Puma and phased restarts we have had a couple of errors:

`No connection pool with 'primary' found`
`already initialized constant X`

While doing more testing we discovered things are not working as if classes were re-loaded but initializers did not run on start.
We have tried switching from Phased to Hot restarts but eventually the only thing that proved working was shutting down the Puma process and starting fresh.

App details: Puma 6, Rails 5.0, Ruby 2.4

**Comment by dentarg** [comment]
Have you seen https://github.com/puma/puma/blob/master/docs/restart.md#phased-restart? ([permalink](https://github.com/puma/puma/blob/cdb23413ae375d2bf2053fe82e5a50f29bb28fd2/docs/restart.md#phased-restart))

Maybe it could give you some cluses. I'm thinking of

> To support upgrading the application that Puma is serving, ensure `prune_bundler` is enabled and that `preload_app!` is disabled

Otherwise there's very little info to go on here to help... but yeah, start fresh is the most reliable way :)

(Also, consider updating Ruby ;-) and Rails)
